src/event/ngx_event_openssl.c

```diff
@@ -3826,6 +3822,10 @@ ngx_ssl_shutdown(ngx_connection_t *c)
     }
 #endif
 
+        if (sslerr == SSL_ERROR_ZERO_RETURN || ERR_peek_error() == 0) {
+            goto done;
```

Comment:
Some following logic being removed:

```
#if (NGX_SSL && NGX_SSL_ASYNC)
            if (c->async_enable) {
                /* Ignore errors from ngx_ssl_async_process_fds as
                   we want to carry on and close the SSL connection
                   anyway. */
                ngx_ssl_async_process_fds(c);
                if (ngx_del_async_conn) {
                    if (c->num_async_fds) {
                        ngx_del_async_conn(c, NGX_DISABLE_EVENT);
                        c->num_async_fds--;
                    }
                }
                ngx_del_conn(c, NGX_DISABLE_EVENT);
            }
#endif
```

---

> In "if (sslerr == SSL_ERROR_ZERO_RETURN || ERR_peek_error() == 0)" branch, no need "ngx_del_async_conn", because it must be invoked before, so delete the code.

这里先别删除里面的清理逻辑，我看原代码也没有删除。可能比如n != -1的时候，这里是不是还能走到。不过具体我也不清楚，最好先保留了。如果有异议建议你在原来项目上开个issue询问下 https://github.com/intel/asynch_mode_nginx

Reply:
cc @oyaya 请一个专家帮忙看看这段能删除不。

Reply:
我来解释一下为什么要删除if (sslerr == SSL_ERROR_ZERO_RETURN || ERR_peek_error() == 0)里面的清理逻辑。为了叙述的方便，我把此处的这个分支代码标记为分支2，因为这个分支2的代码移到了if (sslerr == SSL_ERROR_WANT_ASYNC) 分支（这个分支代码标记为分支1）代码的后面了，仔细看if (sslerr == SSL_ERROR_WANT_ASYNC) 分支1的代码，是先判断if (c->async_enable)成立，再判断if (sslerr == SSL_ERROR_WANT_ASYNC) 成立就直接返回NGX_AGAIN了，如果不成立就会进行ngx_del_async_conn逻辑的清理，如果分支2的这段代码不删除，那么就会和分支1中的重复清理了，所以走到分支2时那段清理逻辑必然已经执行了，所以需要删除分支2的清理逻辑，否则会被清理两次。

> #if (NGX_SSL && NGX_SSL_ASYNC)
		if (c->async_enable) {                                                  **// 分支1**
			if (sslerr == SSL_ERROR_WANT_ASYNC) {
				c->async->handler = ngx_ssl_shutdown_async_handler;
				c->read->saved_handler = ngx_ssl_shutdown_handler;
				c->read->handler = ngx_ssl_empty_handler;
				c->write->handler = ngx_ssl_shutdown_handler;

				ngx_log_debug1(NGX_LOG_DEBUG_EVENT, c->log, 0,
							   "SSL ASYNC WANT recieved: \"%s\"", __func__);

				/* Ignore errors from ngx_ssl_async_process_fds as
				   we want to carry on anyway */
				ngx_ssl_async_process_fds(c);
				return NGX_AGAIN;
			}

			/* Ignore errors from ngx_ssl_async_process_fds as
			   we want to carry on and close the SSL connection
			   anyway. */
			ngx_ssl_async_process_fds(c);
			if (ngx_del_async_conn) {
				if (c->num_async_fds) {
					ngx_del_async_conn(c, NGX_DISABLE_EVENT);                **// 分支1中的清理逻辑**
					c->num_async_fds--;
				}
			}
			ngx_del_conn(c, NGX_DISABLE_EVENT);
		}
#endif

        if (sslerr == SSL_ERROR_ZERO_RETURN || ERR_peek_error() == 0) {      **// 分支2**
            goto done;                        **// 分支2不需要这个清理逻辑了，因为能走到此处，那么在分支1中必然已经清理了**
        }

        err = (sslerr == SSL_ERROR_SYSCALL) ? ngx_errno : 0;

        ngx_ssl_connection_error(c, sslerr, err, "SSL_shutdown() failed");

        break;
    }

Reply:
这个c->read->active的判断我之前也注意到了，但是我并没有修改进去的原因如下：
1.这个github(https://github.com/intel/asynch_mode_nginx)其实是intel在nginx的基础上实现的异步模式，并不代表nginx官方实现，然后咱的tengine也是在nginx的基础上实现的异步模式，和intel的实现可能并不是完全一致的，所以我并没有直接修改进来。
2.我认为既然走到了shutdown流程那么epoll的fd必然已经创建了，所以c->read->active必然成立，感觉这个判断有点多余，退一万步来讲，就算c->read->active不成立，但是还是调用ngx_del_conn也并不会产生不良影响。
3.当前PR要解决的问题和这个c->read->active的判断并没有关系，并且tengine原先所有调用ngx_del_conn的地方并没有去判断c->read->active，所以如果要修改进来还需要经过完整的验证才行。但是目前这个PR要解决的问题对我来说非常紧急和重要，所以我恳求如果这个PR没有引入额外的问题的话，能否尽快帮忙收了。
4.争对是否要进行c->read->active判断，我会接下来把所有使用ngx_del_conn的地方都替换掉，再完整验证一下，然后新提一个PR。
看看这样处理是否可行？

Reply:
可以。我觉得也是这样，最好在提一个新的。这个先merge ty